File: 7_vigenere_cipher/vigenere.py
# Vigenere Cipher in Python

alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
stringSecret = "Key"  # Secret key used
secretKeyIndexes = []  # Indexes of letters in string secret
secretKeyLength = len(stringSecret)  # length of the string secret


def operation(secret_key, input, decrypt):
    tempCounter, output, allCapsInput = 0, "", input.upper()
    global secretKeyIndexes, secretKeyLength

    if len(secretKeyIndexes) == 0: # Convert Secret key to a set of indexes
        secretKeyIndexes = [alphabet.find(char) for char in secret_key.upper()]

    for char in allCapsInput:
        index = alphabet.find(char)
        if index == -1:
            output += char
        else:
            key = secretKeyIndexes[tempCounter % secretKeyLength]
            if decrypt:
                key = key * -1
            tempCounter +=1
            output += alphabet[(26 + index + key) % 26]
    return output

# Function expressions to make stuff easier
encrypt = lambda text : operation(stringSecret, text, False)
decrypt = lambda text : operation(stringSecret, text, True)

plaintext = "New York!"
ciphertext = encrypt(plaintext) # "XIU ISPU!"
decryptedtext = decrypt(ciphertext) # "NEW YORK!"
print(ciphertext, decryptedtext)

File: 7_vigenere_cipher/vigenere.js
// Vigenere Cipher in JavaScript

const alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("");
let stringSecret = "Key";  // Secret key used
let secretKeyIndexes = [];  // Indexes of letters in string secret
var secretKeyLength = stringSecret.length;  // length of the string secret


function operation(secretKey, input, decrypt) {
    let allCapsInput = `${input}`.toUpperCase().split("");
    let tempCounter = 0; // Temporary Counter
    if (secretKeyIndexes.length == 0) { // Convert secret key to set of indexes
        for (char of secretKey.toUpperCase()) {
            secretKeyIndexes.push(alphabet.indexOf(char));
        }
    }

    return allCapsInput.map(
        (char) => {
            let index = alphabet.indexOf(char);
            if (index == null || index == -1) return char;
            else {
                let key = secretKeyIndexes[tempCounter % secretKeyLength];
                tempCounter += 1;
                if (decrypt) key = key * -1;

                return alphabet[(26 + index + key) % 26];
            }
        }).join("");
}

// Function expressions to make stuff easier
const encrypt = text => operation(stringSecret, text, false);
const decrypt = text => operation(stringSecret, text, true);

let plaintext = "New York!";
let ciphertext = encrypt(plaintext); // "XIU ISPU!"
let decryptedtext = decrypt(ciphertext); // "NEW YORK!"
console.log(ciphertext, decryptedtext);